### PR TITLE
Baby branch

### PR DESCRIPTION
**Pull request by aronnoordam**

### CODE REVIEW

File: gmsh_utils/math_utils.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+        target_normal = np.array([0, 0, 1])
+
+        # Calculate the rotation vector
+        rotation_vector = np.cross(polygon_normal, target_normal)
+
+        # Calculate sin and cos of the angle between the polygon normal and the target normal
+        sin_theta = np.linalg.norm(rotation_vector)
+        cos_theta = np.dot(polygon_normal, target_normal)
+
+        # Normalize the rotation vector
```
Comment (brunozc) [inline gmsh_utils/math_utils.py:36]: normalisation is rotation_vector = rotation_vector / np.linalg.norm(rotation_vector).

File: gmsh_utils/math_utils.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+                            [-rotation_vector[1], rotation_vector[0], 0]])
+
+        # Rotation matrix using the Rodrigues' rotation formula
+        rotation_matrix: npt.NDArray[np.float64] = (np.eye(3) + sin_theta * v_cross +
+                                                    (1 - cos_theta) * np.dot(v_cross, v_cross))
+
+        return rotation_matrix
+
+    @staticmethod
+    def is_point_in_polygon(point: Sequence[float], polygon_vertices: Sequence[Sequence[float]], eps: float = 1e-8) \
```
Comment (brunozc) [inline gmsh_utils/math_utils.py:52]: tol move to globals file?

File: gmsh_utils/math_utils.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    def is_point_on_plane(point: npt.NDArray[np.float64],
+                          plane_point: npt.NDArray[np.float64],
+                          plane_normal: npt.NDArray[np.float64]) -> bool:
+        """
+        Check if a point is on a plane.
+
+        Args:
+            - point (npt.NDArray[np.float64]): Point to check.
+            - plane_point (npt.NDArray[np.float64]): Point on the plane.
+            - plane_normal (npt.NDArray[np.float64]): Normal of the plane.
```
Comment (brunozc) [inline gmsh_utils/math_utils.py:139]:
```suggestion
            - plane_point (npt.NDArray[np.float64]): Points on the plane.
            - plane_normal (npt.NDArray[np.float64]): Vector normal of the plane.
```

File: gmsh_utils/math_utils.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+        # make sure that the vertices are not collinear
+        for i in range(2, len(plane_vertices)):
+            v2 = plane_vertices[i] - plane_vertices[0]
+            normal: npt.NDArray[np.float64] = np.cross(v1, v2)
+
+            # return the normal if it is not zero
+            if not np.allclose(normal, 0):
+                return normal / np.linalg.norm(normal)
+
+        raise ValueError("All plane vertices are collinear.")
```
Comment (brunozc) [inline gmsh_utils/math_utils.py:177]: This is only for 3 nodes? if so, no need for the for loop and this should work: 
    v1 = plane_vertices[1] - plane_vertices[0]
    v2 = plane_vertices[i] - plane_vertices[0]
    normal_vector = np.cross(v1, v2)
    normal_vector_normalized = normal_vector / np.linalg.norm(normal_vector)
    
    if np.all(normal_vector == 0):
       ValueError

File: gmsh_utils/gmsh_IO.py [hunk 5]
```diff
@@ -857,6 +859,68 @@ def add_physical_group(self, name: str, ndim: int, geometry_ids: Sequence[int]):
...
+            # check if all surface coordinates are on the plane, if so add surface id to list
+            if all(MathUtils.is_point_on_plane(point, plane_vertices_array[0], normal_plane)
+                   for point in surface_coordinates):
+                surface_ids_at_plane.append(surface_id)
+
+        return surface_ids_at_plane
+
+    def get_surface_ids_at_polygon(self, polygon_vertices: Sequence[Sequence[float]]) -> List[int]:
+        """
+        Gets surface ids at a bounded plane
```
[the remark this answers is not in the document]
Comment (aronnoordam) [reply gmsh_utils/gmsh_IO.py:898]: this is not what is happening, the Args docstring is incorrect

File: gmsh_utils/math_utils.py (hunk 1 above)
Comment (aronnoordam) [reply gmsh_utils/math_utils.py:36]: non normalised rotation vector = np.cross(polygon_normal, target_normal),

||a × b|| = ||a|| * ||b|| * sin(θ)

||np.cross(polygon_normal, target_normal)|| = ||polygon_normal|| * ||target_normal|| * sin(θ)
If your normals are unit vectors (which they typically should be for surface normals), then ||polygon_normal|| = ||target_normal|| = 1, so it simplifies to:
||np.cross(polygon_normal, target_normal)|| = sin(θ)

in short: sin(θ) == np.linalg.norm(non_normalised_rotation_vector)

File: gmsh_utils/math_utils.py (hunk 3 above)
Comment (aronnoordam) [reply gmsh_utils/math_utils.py:139]: its meant for a singular point, as the function is written now

File: gmsh_utils/math_utils.py (hunk 4 above)
Comment (aronnoordam) [reply gmsh_utils/math_utils.py:177]: no the docstring is wrong, it should work for any polygon

**Review by brunozc — APPROVED** [review]
